diff --git a/vendor/tcl8.6.11/unix/configure b/vendor/tcl8.6.11/unix/configure
index b71cdaad9..1a10dbab1 100755
--- a/vendor/tcl8.6.11/unix/configure
+++ b/vendor/tcl8.6.11/unix/configure
@@ -14091,7 +14091,7 @@ main ()
 {
 extern long timezone;
 	    timezone += 1;
-	    exit (0);
+	    return (0);
   ;
   return 0;
 }
@@ -14157,7 +14157,7 @@ main ()
 {
 extern time_t timezone;
 		timezone += 1;
-		exit (0);
+		return (0);
   ;
   return 0;
 }
