Allow kqueue notifier on Darwin.
--- vendor/tcl/unix/configure.orig	2026-07-01 15:53:21
+++ vendor/tcl/unix/configure	2026-07-01 15:55:27
@@ -9742,7 +9742,7 @@
 fi
 
 done;;
-  xDragonFlyBSD|xFreeBSD|xNetBSD|xOpenBSD)
+  xDragonFlyBSD|xFreeBSD|xNetBSD|xOpenBSD|xDarwin)
 	{ printf '%s\n' "$as_me:${as_lineno-$LINENO}: result: kqueue(2)" >&5
 printf '%s\n' "kqueue(2)" >&6; }
 	# Messy because we want to check if *all* the headers are present, and not
