Don't use strchrnul on macOS < 15.4.

--- vendor/tcl/pkgs/sqlite3.51.0/compat/sqlite3/sqlite3.c.orig	2026-04-10 09:47:22
+++ vendor/tcl/pkgs/sqlite3.51.0/compat/sqlite3/sqlite3.c	2026-04-10 09:55:38
@@ -32269,13 +32269,21 @@ SQLITE_API void sqlite3_str_vappendf(
     if( c!='%' ){
       bufpt = (char *)fmt;
 #if HAVE_STRCHRNUL
-      fmt = strchrnul(fmt, '%');
+#if __APPLE__
+      if (__builtin_available(macOS 15.4, iOS 18.4, tvOS 18.4, watchOS 11.4, *))
 #else
-      fmt = strchr(fmt, '%');
-      if( fmt==0 ){
-        fmt = bufpt + strlen(bufpt);
-      }
+      if (1)
+#endif
+      {
+        fmt = strchrnul(fmt, '%');
+      } else
 #endif
+      {
+        fmt = strchr(fmt, '%');
+        if( fmt==0 ){
+            fmt = bufpt + strlen(bufpt);
+        }
+      }
       sqlite3_str_append(pAccum, bufpt, (int)(fmt - bufpt));
       if( *fmt==0 ) break;
     }
