diff -u src/thread2.6.2/Makefile.in src/thread2.6.2/Makefile.in
--- src/thread2.6.2/Makefile.in	2005-07-28 17:15:51.000000000 +0900
+++ src/thread2.6.2/Makefile.in	2005-08-04 21:25:44.000000000 +0900
@@ -70,9 +70,9 @@
 exec_prefix	= @exec_prefix@
 
 bindir		= @bindir@
-libdir		= @libdir@
+libdir		= @prefix@/share/darwinports/Tcl
 datadir		= @datadir@
-mandir		= @mandir@
+mandir		= @prefix@/share/man
 includedir	= @includedir@
 
 DESTDIR		=
diff -u src/thread2.6.2/configure src/thread2.6.2.mine/configure
--- src/thread2.6.2/configure	2005-07-28 17:20:59.000000000 +0900
+++ src/thread2.6.2.mine/configure	2005-08-04 21:12:07.000000000 +0900
@@ -6489,13 +6478,11 @@
                     list=""
                     ;;
             esac
-	    # Check order: pkg --prefix location, Tcl's --prefix location,
+	    # Check order:
 	    # directory of tclConfig.sh, and Tcl source directory.
 	    # Looking in the source dir is not ideal, but OK.
 
-	    eval "temp_includedir=${includedir}"
 	    list="$list \
-		`ls -d ${temp_includedir}        2>/dev/null` \
 		`ls -d ${TCL_PREFIX}/include     2>/dev/null` \
 		`ls -d ${TCL_BIN_DIR}/../include 2>/dev/null` \
 		`ls -d ${TCL_SRC_DIR}/generic    2>/dev/null`"
